rfcs/text/0000-basic-typed-contracts.md

```diff
@@ -0,0 +1,249 @@
+our `(->/c positive? positive?)` the type is
+
+    (Contract (-> {2} Positive-Real Real {3})
+              (-> {1} Real Positive-Real {4}))
+
+In the output type of the contract, here a function type, the domain
+type is `Real` (marked `{1}`). This is because after applying the
+contract to a function, calling the result with a value will apply the
+domain contract (with input type `Real`) to that value. If the value
+passes the domain contract, we know it is a `Positive-Real` by the
+domain contract's output type, which is reflected in the input type's
+domain contract (marked `{2}`). This means that a function we apply
+this contract to has to accept `Positive-Real` values.
+
+In the input type of the contract, the range of the type is `Real`
+(marked `{3}`) because applying the contract to a function is only
+allowed if that function returns a value that can flow through the
+range contract (whose input type is `Real`). Finally, in the output
+type of the contract the range type is `Positive-Real` (marked `{4}`)
+because that's the output type of the range contract. This means that
+if we apply this contract to a function, then the resuling function
+returns at least a `Positive-Real` if it returns at all.
+
+# Reference-level explanation
+
+## Interaction with other features
+
+As far as I can tell contracts are mostly self-contained. Some
+exceptions:
+
+  - Predicates are subtypes of the `FlatContract` type.
+
+  - Values of `FlatContract` type are not functions (`FlatContract`
+    currently corresponds to Racket's `flat-contract?` which is `#t`
+    for e.g. numbers)
+
+  - Applying a function contract (with the current type rule for
+    contract application) combines the range type of the function
+    contract's output type with the range type of the function that
+    will be monitored. In the code this is called "pairwise
+    intersection" and in the thesis this is the `comb` operator.
+
+As far as subtyping goes, `(Contract S T)` and `(FlatContract S T)` is
+contravariant in `S` and covariant in `T`.
+
+## How the feature would be implemented
+
+The MS thesis linked up top gives type rules formalizing what I walked
+through in the guide-level explanation. It also discuss some of the
+implementation techniques used.
+
+(The current implementation may have some support for this part of the
+RFC, too.)
+
+## Corner cases
+
+Because I haven't worked out how to compile a contract type to a
+contract for typed/untyped interaction, typed code can't let untyped
+code use a value of contract type. At the moment I believe the current
+implementation raises an error (maybe an internal one?) when providing
+a value of contract type because I haven't defined a way to compile
+those types to contracts.
+
+I haven't properly looked at how these contract types interact with
+polymorphic types. Currently the only way to get primitive contracts
+with such a type are a predicates on values of some polymorphic type,
+e.g. a function like `null?`.
```

Comment:
For now lets say both things aren't supported, and change the PR to give useful error messages?

Reply:
Updated the RFC!